---
title: "RCHOP Signature"
author: "LIPINSKI Boris"
date: '2022-04-01'
output:
  rmdformats::readthedown:
    code_folding: hide
    fig.width: 9
    fig.height: 7
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
path <- "/home/boris/Bureau/Flinovo/"
path.meta <- paste0(path,"result/analyse_meta/")
path.patient <- paste0(path,"result/analyse_patient/")
source(file = paste0(path,"fonction/library.R"))
source(file = paste0(path,"fonction/workflow.R"))
source(file = paste0(path,"fonction/aggregate.R"))
work="/home/boris/Documents/lipinskib/boris"
source(paste0(work,"/ATACseq/script/Functions_ATAC_v4.R"))
library(pheatmap)
library(scater)
annotations <- read.csv(paste0(path,"document/annotation.csv"))
FC <- 1.109569 #log2(FC) = 0.15 ; FC = 2^0.15 = 1.109569
# means that the expression of that gene is increased in WT relative to KO by a multiplicative factor of 2^log(FC) ≈ FC

# EdgeR : Modèle multivarié en pseudo-bulk
condition <- "Post-greffe"
phénotype <- "B" ;
load(file = paste0(path.meta,"All_",condition,"_",phénotype,".RData")) ; singlet <- noig(singlet) ; DefaultAssay(singlet) <- "RNA" ; Idents(singlet) <- "Condition"

sc <- PB(singlet)
sc.EdgeR <- RNAseq(sc[["cluster_counts"]], sc[["cluster_metadata"]], sc[["cluster_metadata"]]$Condition)
sc.EdgeR <- sc.EdgeR[,which(!sc.EdgeR$samples$sample_id == "FL05G0330RCHOP")]
sc.EdgeR <- EdgeR(sc.EdgeR, model.matrix(~ Sample+Condition, data=sc.EdgeR$samples), FC=log(FC), lfc=1.11)
signature.EdgeR.PB.RCHOP <- rownames(sc.EdgeR[["tr"]]$table[sc.EdgeR[["tr"]]$table$FDR < 0.05,])
#Dimension de la matrice de comptage : `r dim(counts)[1]` gènes exprimés x `r dim(counts)[2]` souris.
```

# Pseudo-Bulk 
## Clustering
```{r echo=F, warning=F, fig.width=9, fig.height=8}
Clustering(sc.EdgeR)
```

## PCA
```{r echo=F}
PCA_edger(sc.EdgeR)
```

## Liste des gènes différentiellement exprimés 
```{r echo=F}
result.RCHOP <- signif(as.data.frame(sc.EdgeR[['result']][,c("logFC", "PValue", "FDR")]),3)
DT::datatable(result.RCHOP, class = 'cell-border stripe')
signature.RCHOP <- rownames(result.RCHOP)
```

## Top20 des genes les plus différentiellement exprimés
```{r echo=F, warning=F, fig.width=10, fig.height=8}
Top20(sc.EdgeR,30)
```

## Volcano plot : FDRvsFC 
```{r echo=F, warning=F, message=F}
#& FCvsCPM
Volcano_plot(sc.EdgeR, xscale = 4.5, yscale = 6, FC=log(FC), seuil=log(FC))
```

## Heatmap : Tous les DEGs
```{r echo=F, fig.width=9, fig.height=11}
col <- c("FL02G095Excipient","FL08G0293Excipient", "FL05G0330Excipient","FL06G1206Excipient","FL08G0404Excipient","FL08G0431Excipient","FL09C1164Excipient","FL12C1888Excipient", "FL140304Excipient",
         "FL06G1206RCHOP","FL08G0293RCHOP","FL02G095RCHOP","FL08G0404RCHOP","FL08G0431RCHOP","FL09C1164RCHOP","FL12C1888RCHOP","FL140304RCHOP")
TopHeatmap(sc.EdgeR, FDR=0.1, FC=0.5, col = col, cluster_cols = F)
```

# Single-cell
## PCA
```{r}
Condition <- "Condition"; item1 <- "RCHOP"; item2 <- "Excipient";
Seurat::DimPlot(object = singlet, group.by = "Condition", label.size = 5, pt.size = 1,reduction = "pca", label = T) & Seurat::NoLegend() & 
  xlab(label = paste0("PCA 1 : ", round(Seurat::Stdev(singlet[["pca"]])[1],2), " %")) & ylab(label = paste0("PCA 2 : ", round(Seurat::Stdev(singlet[["pca"]])[2],2), " %"))
```

## Liste des gènes différentiellement exprimés
```{r}
# RNA
singlet <- NormalizeData(singlet, assay = "RNA")
singlet <- ScaleData(singlet, features = rownames(singlet))
response <- FindMarkers(singlet, ident.1 = item1, ident.2 = item2 , logfc.threshold = log2(FC), assay = "RNA", slot = "data")

response <- response %>% rownames_to_column(var="gene") %>% left_join(y = unique(annotations[, c("gene_name", "description")]), by = c("gene" = "gene_name")) %>% as_tibble(response) 
response1 <- signif(as.data.frame(response[,c("p_val", "avg_log2FC", "p_val_adj")]),3)
response <- cbind(response1, response[,c("gene", "pct.1", "pct.2", "description")])
response <- response[,c("gene", "p_val", "avg_log2FC", "pct.1", "pct.2", "p_val_adj", "description")]
colnames(response) <- c("genes", "p_val", "logFC", "pct.1", "pct.2", "FDR", "description")

response[response$logFC > log(FC) & response$FDR < 0.05, "Expression"] <- "Up"
response[response$logFC < -log(FC) & response$FDR < 0.05, "Expression"] <- "Down"
response[response$FDR >= 0.05,  "Expression"] <- "Not Sig"
response[response$logFC > -log(FC) & response$logFC < log(FC), "Expression"] <- "Not Sig"

DT::datatable(response, class = 'cell-border stripe')
```


## Volcano plot : FDRvsFC 
```{r, warning=FALSE}
data <- response
xscale = 1 ; yscale=150 ; seuil=log(FC) ; label=NULL
if(length(unique(data$Expression))==2){color = c("grey","#F8766D")}else{color = c("#619CFF","grey","#F8766D")}
if(is.null(label)){filter <- data %>% filter(-log10(data$FDR)>1.2 & abs(data$logFC)>FC)}else{filter <- data %>% filter(data$genes %in% label & -log10(data$FDR)>1.2 & abs(data$logFC)>FC)}

ggplot(data) +
  geom_point(aes(x = logFC, y = -log10(FDR), colour = Expression)) + 
  ggrepel::geom_label_repel(data=filter, aes(label=genes, x = logFC, y = -log10(FDR))) +
  ggtitle("\nVolcano plot WT vs KO : FDR < 0.05 & logFC > 1.2\n") + xlab("log2 FC") + ylab("-log10 FDR") +
  scale_x_continuous(limits = c(-xscale,xscale)) + scale_y_continuous(limits = c(0,yscale)) + scale_color_manual(values = color) + 
  geom_hline(yintercept=1.30103, col="black", linetype="dashed") + geom_vline(xintercept=c(-seuil,seuil), col="black", linetype="dashed") +
  theme_classic() + theme(plot.title = element_text(size = rel(1.5), hjust = 0.5), axis.title = element_text(size = rel(1))) 
```

## Heatmap : Tous les DEGs
```{r, warning=FALSE, fig.width=9, fig.height=12}
DoHeatmap(singlet, features = response$genes, group.by = "Condition", slot = "scale.data", assay = "integrated")
```

# Intersection
## Venn Diagram
```{r, echo=FALSE, message=FALSE, error=FALSE}
library(ggVennDiagram)
x <- list(PseudoBulk = signature.RCHOP, SingleCell = response$genes)
ggVennDiagram(x, category.names = c("PseudoBulk","SingleCell"), label_alpha = 1, label = "count") +
  scale_fill_gradient(low="white",high = "white") + theme(legend.position = "none") + ggtitle("DE list intersection") + theme(plot.title = element_text(hjust = 0.5))
```

## Liste
```{r}
intersect(response$genes, signature.RCHOP)
```